### PR TITLE
Add new test for createDataChannel

### PR DESCRIPTION
**Pull request by soareschen**
Adds type coverage annotation and a few new tests for `createDataChannel`.

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240303275)

# Firefox (nightly) #
Testing web-platform-tests at revision cc3c579b9e12864cd4a49bd1fb1a1b7cd51f802e
Using browser at version BuildID 20170606100219; SourceStamp 58ce95bc58ce4ba200413c8bed87786dccf3d105
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6176/webrtc/RTCPeerConnection-createDataChannel.html">/webrtc/RTCPeerConnection-createDataChannel.html</a></summary>


|                                       Subtest                                       | Results |                                                                                                                                                  Messages                                                                                                                                                  |
|-------------------------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                     | OK      |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel required arguments`                                              | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with closed connection`                                          | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel defaults`                                                        | FAIL    | `RTCDataChannel is not defined`                                                                                                                                                                                                                                                                            |
| `createDataChannel with label "foo"`                                                | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with label null`                                                 | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with label undefined`                                            | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with label lone surrogate`                                       | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with ordered false`                                              | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with ordered null/undefined`                                     | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with maxPacketLifeTime 0`                                        | FAIL    | `assert_equals: expected (number) 0 but got (undefined) undefined`                                                                                                                                                                                                                                         |
| `createDataChannel with maxRetransmits 0`                                           | FAIL    | `assert_equals: expected (number) 0 but got (undefined) undefined`                                                                                                                                                                                                                                         |
| `createDataChannel with both maxPacketLifeTime and maxRetransmits`                  | FAIL    | `assert_throws: function "() => pc.createDataChannel('', {\n    maxPacketLifeTime: 0,\n    maxRetransmits: 0\n  })" threw object "InvalidParameterError: Both maxPacketLifeTime and maxRetransmits cannot be provided" that is not a DOMException SyntaxError: property "code" is equal to 0, expected 12` |
| `createDataChannel with protocol "foo"`                                             | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with protocol null`                                              | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with protocol undefined`                                         | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with protocol lone surrogate`                                    | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with negotiated true`                                            | FAIL    | `assert_equals: expected (boolean) true but got (undefined) undefined`                                                                                                                                                                                                                                     |
| `createDataChannel with id -1`                                                      | FAIL    | `assert_throws: function "() => pc.createDataChannel('', { id: id })" did not throw`                                                                                                                                                                                                                       |
| `createDataChannel with id 0`                                                       | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with id 1`                                                       | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with id 65534`                                                   | PASS    |                                                                                                                                                                                                                                                                                                            |
| `createDataChannel with id 65535`                                                   | FAIL    | `assert_throws: function "() => pc.createDataChannel('', { id: id })" did not throw`                                                                                                                                                                                                                       |
| `createDataChannel with id 65536`                                                   | FAIL    | `assert_throws: function "() => pc.createDataChannel('', { id: id })" did not throw`                                                                                                                                                                                                                       |
| `createDataChannel with priority "high"`                                            | FAIL    | `assert_equals: expected (string) "high" but got (undefined) undefined`                                                                                                                                                                                                                                    |
| `createDataChannel with invalid priority should throw TypeError`                    | FAIL    | `assert_throws: function "() => pc.createDataChannel('', { priority: 'invalid' })" did not throw`                                                                                                                                                                                                          |
| `createDataChannel with negotiated false and long label should throw TypeError`     | FAIL    | `assert_throws: function "() =>\n    pc.createDataChannel('', {\n      label: longString,\n      negotiated: false\n    })" did not throw`                                                                                                                                                                 |
| `createDataChannel with negotiated false and long protocol should throw TypeError`  | FAIL    | `assert_throws: function "() =>\n    pc.createDataChannel('', {\n      protocol: longString,\n      negotiated: false\n    })" did not throw`                                                                                                                                                              |
| `createDataChannel with negotiated true and long label and protocol should succeed` | PASS    |                                                                                                                                                                                                                                                                                                            |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240303276)

# Sauce (safari) #
Testing web-platform-tests at revision cc3c579b9e12864cd4a49bd1fb1a1b7cd51f802e
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6176/webrtc/RTCPeerConnection-createDataChannel.html">/webrtc/RTCPeerConnection-createDataChannel.html</a></summary>


|                                       Subtest                                       | Results |                 Messages                 |
|-------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                     | OK      |                                          |
| `createDataChannel required arguments`                                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with closed connection`                                          | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel defaults`                                                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with label "foo"`                                                | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with label null`                                                 | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with label undefined`                                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with label lone surrogate`                                       | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with ordered false`                                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with ordered null/undefined`                                     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with maxPacketLifeTime 0`                                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with maxRetransmits 0`                                           | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with both maxPacketLifeTime and maxRetransmits`                  | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with protocol "foo"`                                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with protocol null`                                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with protocol undefined`                                         | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with protocol lone surrogate`                                    | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with negotiated true`                                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with id -1`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with id 0`                                                       | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with id 1`                                                       | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with id 65534`                                                   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with id 65535`                                                   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with id 65536`                                                   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with priority "high"`                                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with invalid priority should throw TypeError`                    | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with negotiated false and long label should throw TypeError`     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with negotiated false and long protocol should throw TypeError`  | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createDataChannel with negotiated true and long label and protocol should succeed` | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240303282)

# Chrome (unstable) #
Testing web-platform-tests at revision cc3c579b9e12864cd4a49bd1fb1a1b7cd51f802e
Using browser at version 60.0.3112.20 dev
Starting 10 test iterations

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240303283)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision cc3c579b9e12864cd4a49bd1fb1a1b7cd51f802e
Using browser at version 14.14393
Starting 10 test iterations

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 1]
```diff
@@ -121,9 +220,98 @@
...
+    () => pc.createDataChannel('', { priority: 'invalid' }));
+}, 'createDataChannel with invalid priority should throw TypeError');
+
+function generateLongString(length) {
+  let str = '';
+  for(let i=0; i<length; i++) {
+    str += 'a';
+  }
+  return str;
+}
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: This is unnecessary, just use `" ".repeat(65536)`

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 2]
```diff
@@ -121,9 +220,98 @@
...
+/*
+  6.2.  createDataChannel
+    6.  If negotiated is false and label is longer than 65535 bytes long,
+        throw a TypeError.
+ */
+test(() => {
+  const pc = new RTCPeerConnection();
+  assert_throws(new TypeError(), () =>
+    pc.createDataChannel('', {
+      label: longString,
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: use `" ".repeat(65536)`

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 3]
```diff
@@ -121,9 +220,98 @@
...
+  6.2.  createDataChannel
+    7.  If negotiated is false and protocol is longer than 65535 bytes long,
+        throw a TypeError.
+ */
+test(() => {
+  const pc = new RTCPeerConnection();
+  const protocol = generateLongString(65536);
+  assert_throws(new TypeError(), () =>
+    pc.createDataChannel('', {
+      protocol: longString,
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: use `" ".repeat(65536)`

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 4]
```diff
@@ -121,9 +220,98 @@
...
+      negotiated: false
+    }));
+}, 'createDataChannel with negotiated false and long protocol should throw TypeError');
+
+test(() => {
+  const pc = new RTCPeerConnection();
+  pc.createDataChannel('', {
+    label: longString,
+    protocol: longString,
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: use `" ".repeat(65536)`

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 5]
```diff
@@ -121,9 +220,98 @@
...
+}, 'createDataChannel with negotiated false and long label should throw TypeError');
+
+/*
+  6.2.  createDataChannel
+    7.  If negotiated is false and protocol is longer than 65535 bytes long,
+        throw a TypeError.
+ */
+test(() => {
+  const pc = new RTCPeerConnection();
+  const protocol = generateLongString(65536);
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: Delete this line, it's not used

**Review by rwaldron — CHANGES_REQUESTED** [review]

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 6]
```diff
@@ -6,19 +6,56 @@
 <script>
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: Above this line, change the title to "RTCPeerConnection.prototype.createDataChannel"

File: webrtc/RTCPeerConnection-createDataChannel.html (hunk 1 above)
Comment (soareschen) [inline webrtc/RTCPeerConnection-createDataChannel.html]: Ah nice. Didn't know about that.

**Comment by soareschen** [comment]
Fixed the feedback. I also restyled the old code a bit to make it look more consistent with the new style I adopt.

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 7]
```diff
@@ -1,26 +1,63 @@
 <!doctype html>
 <meta charset=utf-8>
-<title>RTCPeerConnection createDataChannel method</title>
+<title>RTCPeerConnection.prototype.createDataChannel</title>
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: Thank you!

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 8]
```diff
@@ -1,26 +1,63 @@
...
+ */
+
 test(() => {
-  const pc = new RTCPeerConnection;
+  const pc = new RTCPeerConnection();
   assert_equals(pc.createDataChannel.length, 1);
-  assert_throws(new TypeError, () => pc.createDataChannel());
-}, 'createDataChannel required arguments');
+  assert_throws(new TypeError(), () => pc.createDataChannel());
+}, 'createDataChannel with no argument should throw TypeError');
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: Excellent!!

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 9]
```diff
@@ -94,36 +175,132 @@
...
+      10. If id is equal to 65535, which is greater than the maximum allowed ID
+          of 65534 but still qualifies as an unsigned short, throw a TypeError.
+ */
+for (const id of [0, 1, 65534]) {
+  test(() => {
+    const pc = new RTCPeerConnection();
+    const channel = pc.createDataChannel('', { id });
+    assert_equals(channel.id, id);
+  }, `createDataChannel with id ${id} should succeed`);
+}
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: 👍

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 10]
```diff
@@ -94,36 +175,132 @@
...
+
+for (const id of [-1, 65535, 65536]) {
+  test(() => {
+    const pc = new RTCPeerConnection();
+    assert_throws(new TypeError(), () => pc.createDataChannel('', { id }));
+  }, `createDataChannel with id ${id} should throw TypeError`);
 }
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: 👍

File: webrtc/RTCPeerConnection-createDataChannel.html [hunk 11]
```diff
@@ -94,36 +175,132 @@
...
+        If no available ID could be generated, or if the value of the id member
+        of the dictionary is taken by an existing RTCDataChannel , throw a
+        ResourceInUse exception.
+
+  Untestable
+  6.2.  createDataChannel
+    9.  If an attribute, either maxPacketLifeTime or maxRetransmits , has been
+        set to indicate unreliable mode, and that value exceeds the maximum
+        value supported by the user agent, the value must be set to the user
+        agents maximum value.
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-createDataChannel.html]: Once again, thank you for providing this TODO section :)

**Review by rwaldron — APPROVED** [review]

**Pull request merged**